components/raftstore-v2/src/bootstrap.rs

```diff
@@ -167,75 +255,4 @@ impl<'a, ER: RaftEngine> Bootstrap<'a, ER> {
         box_try!(self.engine.consume(&mut wb, true));
         Ok(())
     }
-
-    fn inner_bootstrap_first_region(
```

Comment:
Why move them around?

Reply:
Which part are you referring to? This function is removed because @Connor1996 suggested it and I agree it looks better. Or are you referring to the ordering between public function and private sub-routines? My habit is to read public functions first, private methods later (just like C++ code). If you have a better reason suggesting the opposite I'm willing to change it back.

Reply:
I refer to the ordering of functions. Everyone has their habits. To me, related functions should be put together no matter what access level they are. However, I don't think it's a good idea to reorder them again after they are committed to the code base unless for common sense.

Reply:
`inner_bootstrap_first_region` is extracted without any sense, it can be just placed into `bootstrap_first_region`. Moreover, `inner_bootstrap_first_region` disturbs the ordering for reading code, you need to read back and forth. Now, it's clean and neat.

Reply:
So we can all agree related functions should be grouped together. As for the ordering, I believe it's favorable to arrange code as we read them. In the old days this is mostly impossible because of name resolution constraints, but with Rust I think we should arrange things naturally whenever we can (which is top down, reading flow, in contrary to writing flow).

Reply:
I think reading several small functions with meaningful names is a lot easier than reading a long functions with a lot of comments. But this function is not too long to be split, so I'm OK with the change.

Reply:
> I think we should arrange things naturally whenever we can

I don't think so as there is no consensus on what's natural. For example, for peer struct, what's the natural way to order all those the functions?

And when reading using vscode/vim/ides, there is no much difference between top/down or jumping using builtin support tools.

I consider reordering functions evil as it's hard to review. If the suggestion is proposed the first time PR is sent, it's acceptable.

Reply:
> If the suggestion is proposed the first time PR is sent, it's acceptable.

Clearly we need some basic style guides here, at least some sense of what is considered more readable. I can relate that ordering is not very important. Especially when the member functions keep growing and get further intertwined.

> I think reading several small functions with meaningful names is a lot easier than reading a long functions with a lot of comments.

However, I want to fight a little on this (even though it's not the point of this PR). Splitting a function creates cognitive load for the reader. When reading the code, one has to keep the control flow context inside the (human) memory, the same way that a computer would keep stack variables around. So I want to make sure we can all agree that function calling depth should be kept low if possible.

Furthermore for raftstore code, most functions mutate some internal states (let it be engine data, memory states, network states). Creating sub-routines make the state mutation hard to track.

Both of these two points can't be addressed with IDE. And since we are rebuilding the raftstore code, I think it makes sense to push for this and make the code easier to read than v1.

For reference, I love this post and some related HN discussions: https://hn.algolia.com/?dateRange=all&page=0&prefix=true&query=inlined%20code&sort=byPopularity&type=story

Reply:
> I can relate that ordering is not very important.

Yes, to me it's not important. If the comment is sent back in the time, I will happily change it. The most important thing is:
1. Don't randomly change the order of committed code as it's hard to read and it causes trouble of future cherry-picks of potential bugs;
2. Functions with related purpose should be put together to ease context switch. That's the purpose why v2 has `operations` module to split code by operations.

> Splitting a function creates cognitive load for the reader. 

Note the prerequisite is **small functions with meaningful names**. As long as the function name makes sense, all we need to look to is whether the implementation represents its name. So there will be no such requirement to keep tracking variables. For example:
```
if store.needs_continue_bootstrapping() {
    store.bootstrap();
}
```
One can easily know what the flow is by name. Though I'm not suggesting the original naming is precise and correct, just suggesting reusing codes is not the only purpose of extracting functions.

Reply:
Yes, it's a good point. Though many would argue that using function as a way to "group code" isn't really better than just writing some comments. Because to understand `needs_continue_bootstrapping`, you still need to do a context switch. It's against human's nature to put the questions aside and take the function name as its implementation.

In our situation, I am advising strongly on calling depth and state mutations. Good function names can only do as much. And take a second thought before doing a refactor. Veteran devs such as yourself sometimes just get too comfortable with sub-routines and treating them like "common sense", meanwhile for newcomer it would take a long while to get around.